### PR TITLE
[Jia Zheng] Duke Increments

### PR DESCRIPTION
**Pull request by ljiazh3ng**
Level 1-6 of Duke assignment Week 2

### CODE REVIEW

File: src/main/java/DukeException.java [hunk 1]
```diff
@@ -0,0 +1,10 @@
+public class DukeException extends Exception {
```
Comment (don-tay) [inline src/main/java/DukeException.java]: Good to define a new Exception class. Keep it up!

File: src/main/java/Duke.java [hunk 2]
```diff
@@ -1,10 +1,127 @@
...
+            if (!parseMessage(scanner.nextLine(), tasks)) {
+                break;
+            }
+        }
+    }
+
+    /**
+     * Print instruction of the commands in JayZbot.
+     */
+    public void botInstruction() {
```
Comment (don-tay) [inline src/main/java/Duke.java]: Good feature to have instructions for users of the chatbot.

File: src/main/java/Duke.java [hunk 3]
```diff
@@ -1,10 +1,127 @@
...
+            System.out.println(String.format("* %-77s*",' '));
+            System.out.println(String.format("* %-77s*",ex.getMessage()));
+            System.out.println(String.format("* %-77s*",' '));
+            System.out.println(String.format("%80s",' ').replace(' ','*'));
+            return true;
+        }
+    }
+
+    /**
+     * Echo message back in a text box.
```
Comment (don-tay) [inline src/main/java/Duke.java]: Change the word "Echo" to "Echoes" as starting verbs used for javadocs should be singular, present tense.

File: src/main/java/Duke.java [hunk 4]
```diff
@@ -1,10 +1,127 @@
...
+            } else if (message.substring(0,8).equals("deadline")) {
+                tasks.addTask(message, Symbol.D);
+                return true;
+            } else if (message.substring(0,6).equals("delete")) {
+                tasks.deleteTask(message);
+                return true;
+            } else {
+                throw new IllegalArgumentException("☹ OOPS!!! I'm sorry, but I don't know what that means :-(");
+            }
+        } catch (DukeException e) {
```
Comment (don-tay) [inline src/main/java/Duke.java]: Good error handling, order is appropriate.

File: src/main/java/Duke.java [hunk 5]
```diff
@@ -1,10 +1,127 @@
+import java.util.ArrayList;
+import java.util.Scanner;
+
 public class Duke {
+    /**
+     * Process message and deem if user wants to quit.
+     * Print good bye message if user wants to quit.
```
Comment (don-tay) [inline src/main/java/Duke.java]: Use singular, present tense for verbs describing what the method does in Javadocs

File: src/main/java/Symbol.java [hunk 6]
```diff
@@ -0,0 +1,3 @@
+public enum Symbol {
```
Comment (don-tay) [inline src/main/java/Symbol.java]: Well defined enumeration. Appropriately used. Good work!

File: src/main/java/Task.java [hunk 7]
```diff
@@ -0,0 +1,25 @@
+public class Task {
```
Comment (don-tay) [inline src/main/java/Task.java]: Well defined class, perhaps you can consider defining as abstract class instead, for extensibility reasons in the future. You can consider defining the "crossstick" String that were found and separately defined in Event, Deadlines, Todo classes respectively.

File: src/main/java/Task.java [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+    protected String title;
+    protected int index;
+    protected boolean done;
```
Comment (don-tay) [inline src/main/java/Task.java]: Variable should be named "isDone" instead to be more verbose.

File: src/main/java/TaskTracker.java [hunk 9]
```diff
@@ -0,0 +1,129 @@
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.TreeMap;
+
+public class TaskTracker {
+    private TreeMap<Integer, Task> tasks;
```
Comment (don-tay) [inline src/main/java/TaskTracker.java]: Great use of appropriate data structures that helps with scalibility. Keep it up!

**Review by don-tay — COMMENTED** [review]
Overall, neat increments with appropriate OOP design for extensibility, and data structure for scalability. Do note the use of singular, present tense for verbs in Jsvadocs.

File: gradlew [hunk 10]
```diff
@@ -0,0 +1,172 @@
+#!/usr/bin/env sh
```
Comment (jiachen247) [inline gradlew:1]: Personally not a fan of committing gradlew into the vcs! Might want to consider adding it to gitignore!

File: gradlew.bat [hunk 11]
```diff
@@ -0,0 +1,84 @@
+@if "%DEBUG%" == "" @echo off
```
Comment (jiachen247) [inline gradlew.bat:1]: ^ see comment above!

File: src/main/java/duke/Duke.java [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+        }
+    }
+
+    /**
+     * Prints a logo of the name Duke.
+     * @param args Set up main function
+     */
+    public static void main(String[] args) {
+        String home = System.getProperty("user.home");
+        Path path = java.nio.file.Paths.get(home, "Desktop", "GitHub", "duke","data","data.txt");
```
Comment (jiachen247) [inline src/main/java/duke/Duke.java:104]: might wanna keep this as generic as possible! Would suggest creating ./data in the pwd.

File: src/main/java/duke/Duke.java [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+    /**
+     * Chat bot function, run infinitely until user types "bye".
+     */
+    public void runChatBot() {
+        this.ui.welcome();
+        boolean isExit = false;
+        while (!isExit) {
+            try {
+                String fullCommand = ui.readCommand();
+                Command c = Parser.parseMessage(fullCommand);
```
Comment (jiachen247) [inline src/main/java/duke/Duke.java:66]: could use more descriptive names! eg `currentCommand`

File: src/main/java/duke/parser/Parser.java [hunk 14]
```diff
@@ -0,0 +1,112 @@
+package duke.parser;
```
Comment (jiachen247) [inline src/main/java/duke/parser/Parser.java:1]: this file is super arcane! Would suggest being more declarative and breaking them up into smaller functions!

File: text-ui-test/ACTUAL.TXT [hunk 15]
```diff
@@ -0,0 +1,42 @@
+Hello! I'm JayZ-Bot
```
Comment (jiachen247) [inline text-ui-test/ACTUAL.TXT:1]: git ignore this pls!!!!! super bad practice to commit generated files.

**Review by jiachen247 — CHANGES_REQUESTED** [review]
over all LGTM! good use of OOP!

File: src/main/java/duke/Duke.java (hunk 13 above)
Comment (vivienlee01) [reply src/main/java/duke/Duke.java:66]: isExit sounds a bit confusing, perhaps hasExit might be better

File: src/main/java/duke/Duke.java (hunk 12 above)
Comment (vivienlee01) [inline src/main/java/duke/Duke.java:104]: code line seems a bit long, maybe you might want to consider just having ur home and filename var

File: src/main/java/duke/command/AddCommand.java [hunk 16]
```diff
@@ -0,0 +1,42 @@
+package duke.command;
+
+import duke.storage.Storage;
+import duke.task.Task;
+import duke.task.TaskList;
+import duke.ui.Ui;
+
+public class AddCommand extends Command {
+    private final Task task;
+    private final String success = "Success";
```
Comment (vivienlee01) [inline src/main/java/duke/command/AddCommand.java]: I'm not sure if this is a constant, if it is just a reminder to put it in caps!

Comment (vivienlee01) [reply src/main/java/duke/command/AddCommand.java]: Overall really good and clean code!

**Comment by ghost** [comment]
Students made good comments and you might refer to them to improve quality of your codes :)
Overall, well written and structured, but still some parts of codes need to be improved :)
Keep it up!

**Pull request closed without merging**